# 1.1 Conceitos básicos e Aritmética de Ponto Flutuante

## Introdução
De modo geral, a resolução de um problema de engenharia passa necessariamente por 3 etapas distintas
1. observação e entendimento do(s) fenômeno(s) envolvido(s);
2. construção de um modelo matemático que represente o problema;
3. a resolução do modelo matemático por métodos numéricos ou analíticos.

As duas primeiras etapas consistem na **fase de modelagem** e a terceira é a **fase de resolução**. Tanto na fase de modelagem quanto na fase de resolução podem ocorrer erros, mesmo que esses erros não tornem a resposta obtida inválida ou inútil. 

Os erros na fase de modelagem podem ser decorrentes de medições imprecisas, erros de anotação, simplificações e deduções equivocadas. Os erros na fase de resolução ocorrem devido a erros de cálculo, escolha de um método ineficiente, aproximações ou processamento numérico. Dizemos que métodos analíticos de resolução geram soluções exatas e métodos numéricos geram soluções aproximadas. 

Nem sempre é fácil ou possível obter a solução analítica de um modelo matemático, ainda que simples, por exemplo:

1. $$x^3-30x+2552=0 $$ <br>

2. $$cos(x)=x $$ <br>

3. $$xe^x=10 $$ <br>

4. $$\begin{cases} y'+xy = e^{-y}\\
                    y(0) =2
                    \end{cases} $$ <br>
          
5. $$ \int^2_1 e^{-x^2} dx $$ <br>

A maioria dos problemas envolvendo fenômenos reais produzem modelos matemáticos cuja solução analítica é difícil (ou impossível) de obter, mesmo quando provamos que a solução existe. Nesses casos usamos **métodos numéricos**, os quais fornecem **soluções aproximadas**, que apesar de, em geral, serem diferentes da solução exata, podem ser suficientemete próximas para que sejam úteis em aplicações práticas.  





### EXEMPLO 1###
Uma esfera de madeira (densidade relativa $\mu_e =0,638$ e raio $r=10 cm$) é colocada a flutuar sobre a água (densidade relativa $\mu=1$), conforme figura. Qual é a altura $d$ da esfera que ficará submersa? 

![esfera.png](https://github.com/tiagoburiol/NUMETHODS/raw/master/1_INTRODUCAO_ERROS/imagens/esfera.png
)

### Solução###
Inicialmente, observemos que, pelo *Princípio de Arquimedes* o peso da esfera deve ser equilibrado com o peso do volume de água deslocada 

$$ M_e g = M_a g $$

$$ \mu_e \frac{4 \pi r^3}{3} = \mu_a \int_0^d \pi \left[r^2-(x-r)^2) \right]dx $$

$$ \mu_e \frac{4 \pi r^3}{3} = \mu_a \frac{\pi d^2 (3r-d)}{3}$$

$$ 4 \mu_e r^3 = \mu_a d^2 (3r-d)$$

substituindo os valores obtemos 

$$ d^3-30d^2+2552=0$$.

A solução analítica para esse problema pode ser encontrada mas é um tanto complicada. No entanto, com uma rotina podemos encontrar uma aproximação numericamente. 

Sabemos que a solução é um numero entre 0 e 20, então calculamos os valores da função $ f(d) = d^3-30d^2+2552 $ para valores entre 0 e 20 obtendo:


In [4]:
for d in range(20):
    print (d, d**3-30*d**2+2552)

0 2552
1 2523
2 2440
3 2309
4 2136
5 1927
6 1688
7 1425
8 1144
9 851
10 552
11 253
12 -40
13 -321
14 -584
15 -823
16 -1032
17 -1205
18 -1336
19 -1419


Observemos que a função moda de sinal entre $d=11$ e $d=12$, então podemos refinar a aproximação gerando alguns valores entre 11 e 12.

In [6]:
for x in range(10):
    d = 11+x*0.1
    print (d, d**3-30*d**2+2552)

11.0 253.0
11.1 223.33100000000013
11.2 193.72800000000007
11.3 164.19700000000012
11.4 134.7439999999997
11.5 105.375
11.6 76.09599999999955
11.7 46.912999999999556
11.8 17.831999999999425
11.9 -11.141000000000076


Observemos que a função muda de sinal entre $d=11.8$ e $d=11.9$, então podemos refinar a aproximação gerando alguns valores.

In [7]:
for x in range(10):
    d = 11.8+x*0.01
    print (d, d**3-30*d**2+2552)

11.8 17.831999999999425
11.81 14.929740999999922
11.82 12.028567999999723
11.83 9.128486999999495
11.84 6.229503999999906
11.850000000000001 3.3316249999998035
11.860000000000001 0.43485599999894475
11.870000000000001 -2.4607969999997295
11.88 -5.355328000000554
11.89 -8.248731000000134


Assim, podemos tomar como uma solução aproximada o ponto médio entre 11.86 e 11.87 que resolta em $d=11.865 cm$ com erro inferior a 0.005.

### EXEMPLO 2

Um corpo em queda livre pode ser descrito pelo modelo matemático

$$ \frac{dv}{dt} = \frac{F}{m}$$

derivado da 2ª Lei de Newton $F=ma$. A força #F# atuante sobre o corpo composta pela soma de duas forças de mesma direção e sentidos opostos, a força gravitacional para baixo ($F_D$) e a força da resistência do ar para cima ($F_U$). Então

$$ \frac{dv}{dt} = \frac{(F_D+F_U)}{m}$$

como $F_D = mg$, com $m$ massa e $g$ aceleração da gravidade, e $F_U=-cv$, com $c$ sendo uma constante que depende das características do corpo em queda livre e $v$ a velocidade instantânea. Então

$$ \frac{dv}{dt} = \frac{mg-cv}{m}$$

ou

$$ \frac{dv}{dt} = g-\frac{c}{m}v$$

O modelo então consiste em uma equação diferencial cuja solução pode ser obtida analitivamente. Se o corpo em queda livre entrou em movimento a partir do repouso, ou seja, $v=0$ em $t=0$, a sulução analítica é dada por

$$ v(t) = gm(1-e^{(-c/m)t})$$


